### PR TITLE
Add `where` parameter to `union_relations`

### LINKED ISSUE #382: Add optional where condition in dbt-utils.union macro
### Describe the feature
We have a number of source models that are materialized as views that make use of the dbt-utils.union function. We have a number of single tenant databases which are ultimately joined together ready for processing in our staging models. The source models look something like this:

```
{{ config(materialized='view') }}

with all_data as 
(
    {% set live_data = dbt_utils.get_relations_by_pattern('%_SCHEMA_PART_%', 'LIVE_TABLE', '', 'RAW_SCHEMA') %}
    {% set archive_data = dbt_utils.get_relations_by_pattern('%_SCHEMA_PART_%', 'ARCHIVE_TABLE', '', 'RAW_SCHEMA') %}
    {% set all_data = live_data + archive_data %}
    {{ dbt_utils.union_relations(all_data) }}
)

SELECT 
*
FROM all_data

```

This results in queries that run look something like:

```
with all_data as 
(
	(
		select
			cast('RAW_SCHEMA.SCHEMA_PART_DB1.LIVE_TABLE' as varchar) as _dbt_source_relation,
			* -- Fields omitted for brevity
		from RAW_SCHEMA.SCHEMA_PART_DB1.LIVE_TABLE
	)
	
	union all

	(
		select
			cast('RAW_SCHEMA.SCHEMA_PART_DB2.LIVE_TABLE' as varchar) as _dbt_source_relation,
			* -- Fields omitted for brevity
		from RAW_SCHEMA.SCHEMA_PART_DB2.LIVE_TABLE
	)

	union all

	(
		select
			cast('RAW_SCHEMA.SCHEMA_PART_DB1.ARCHIVE_TABLE' as varchar) as _dbt_source_relation,
			* -- Fields omitted for brevity
		from RAW_SCHEMA.SCHEMA_PART_DB1.ARCHIVE_TABLE
	)

	union all
	
	(
		select
			cast('RAW_SCHEMA.SCHEMA_PART_DB2.ARCHIVE_TABLE' as varchar) as _dbt_source_relation,
			* -- Fields omitted for brevity
		from RAW_SCHEMA.SCHEMA_PART_DB2.ARCHIVE_TABLE
	)
)

SELECT 
*
FROM all_data
```

It is necessary for us to filter some of the source data before it goes through to the staging models which we can only do right now by adding the WHERE condition to the final query ie

```
-- Initial CTE omitted for brevity

SELECT 
*
FROM all_data
WHERE active = true
```
It would be useful (and more performant) if we could append these conditions as part of the initial CTE as we get the data from each table so the initial CTE would become something like:

```
with all_data as 
(
	(
		select
			cast('RAW_SCHEMA.SCHEMA_PART_DB1.LIVE_TABLE' as varchar) as _dbt_source_relation,
			* -- Fields omitted for brevity
		from RAW_SCHEMA.SCHEMA_PART_DB1.LIVE_TABLE
		where active = true
	)
	
	union all

	(
		select
			cast('RAW_SCHEMA.SCHEMA_PART_DB2.LIVE_TABLE' as varchar) as _dbt_source_relation,
			* -- Fields omitted for brevity
		from RAW_SCHEMA.SCHEMA_PART_DB2.LIVE_TABLE
		where active = true
	)

	union all

	(
		select
			cast('RAW_SCHEMA.SCHEMA_PART_DB1.ARCHIVE_TABLE' as varchar) as _dbt_source_relation,
			* -- Fields omitted for brevity
		from RAW_SCHEMA.SCHEMA_PART_DB1.ARCHIVE_TABLE
		where active = true
	)

	union all
	
	(
		select
			cast('RAW_SCHEMA.SCHEMA_PART_DB2.ARCHIVE_TABLE' as varchar) as _dbt_source_relation,
			* -- Fields omitted for brevity
		from RAW_SCHEMA.SCHEMA_PART_DB2.ARCHIVE_TABLE
		where active = true
	)
)

SELECT 
*
FROM all_data
```

### Describe alternatives you've considered
Taking dbt-utils.union and making my own macro, however it would be nice to continue to just use the managed versions of this package

### Additional context
I'm fairly new to DBT so it could be that I'm missing a simple trick here, so please point this out to me if that's the case. 

### Who will this benefit?
The example above hopefully makes it clear where the benefit comes from. Being able to filter the source data when making use of the dbt-utils.union macro means that the next stage of the pipeline doesn't have as much data to process

### Are you interested in contributing this feature?
Maybe!

### LINKED ISSUE #464: Add WHERE filter inside the dbt_utils.union macro
### Describe the feature
It will be beneficial to have a possibility to filter large source tables before doing the union macro. It will especially benefit the incremental models. Currently the union is done completely and only afterwards we have a possibility to apply the macro outside the filter, which defies the purpose of the incremental filter. 
In order for the query to work properly, we need to include the WHERE clause into each table's SELECT before the UNION.

E.g. `select * from A where x<y union all select * from B where x<y`  instead of the way it is now  `select * from (select * from A union all select * from B) where x<y`

### Describe alternatives you've considered
I've tried to use the UNION manually without the macros, but it was too complex as the tables don't match exactly. the columns are sorted differently etc.

### Additional context
It's very relevant for Postgres. I'm not sure how other databases optimizators work, I guess it's relevant for the others too.

### Who will this benefit?
It's useful when you have large transactional source tables from different environment and you want to union them together as suggested [in this article](https://discourse.getdbt.com/t/unioning-identically-structured-data-sources/921), but they are too large and require incremental materialization.

### Are you interested in contributing this feature?
Yes, I'm interested. I have already created a modified macro which serves my purpose, but it's not flexible enough to be part of the package. 

P.S. I've also added other features, like creating the relation ref() statement based on source, table and alteration parameters, instead of providing relation directly, union(source= database1, table= table1, environments= [prod, dev, demo]). This help to generate a new unique by appending the environment and adds the environment column for filtering. Could also be consiedered as a feature for a package.

### PR DESCRIPTION
**Pull request by LewisDavies**
Resolves #464, resolves #382

This is a:
- [ ] bug fix PR with no breaking changes — please ensure the base branch is `main`
- [x] new functionality — please ensure the base branch is the latest `dev/` branch
- [ ] a breaking change — please ensure the base branch is the latest `dev/` branch

## Description & motivation
I often want to add basic filters when using `union_relations`, such as excluding obsolete records or low-quality data. Currently, I have to create two models to do so: one to union the data then another to filter it. This PR adds a `where` parameter to `union_relations` so models are filtered _before_ being unioned.

More context & discussion can be found in #464 and #382.

## Checklist
- [x] I have verified that these changes work locally on the following warehouses (Note: it's okay if you do not have access to all warehouses, this helps us understand what has been covered)
    - [ ] BigQuery
    - [ ] Postgres
    - [ ] Redshift
    - [x] Snowflake
- [x] I have updated the README.md (if applicable)
- [x] I have added tests & descriptions to my models (and macros if applicable)
- [x] I have added an entry to CHANGELOG.md

### CODE REVIEW

**Review by dbeatty10 — CHANGES_REQUESTED** [review]
Thanks for submitting this and for including both integration tests plus documentation 🏆 

Thank you also for alerting about the branch names within the merge request template being out-of-date; those are updated now.

## Review
Only feedback:
- I'd like to move the `where` parameter to be at the end of the argument list so we can guarantee no accidental breakage for any consumers.

Is there any reason why it would be beneficial or necessary to be the first keyword argument?

Question:
- It looks like the same filter specified by the `where` parameter will be applied to every relation; I'm assuming you haven't needed unique filters per relation as often?

## Aside
At first, I was wondering if this would conflict or overlap with the [where config](https://docs.getdbt.com/reference/resource-configs/where), but it looks like it is only for  test resources.

File: macros/sql/union.sql [hunk 1]
```diff
@@ -1,8 +1,8 @@
-{%- macro union_relations(relations, column_override=none, include=[], exclude=[], source_column_name='_dbt_source_relation') -%}
-    {{ return(adapter.dispatch('union_relations', 'dbt_utils')(relations, column_override, include, exclude, source_column_name)) }}
+{%- macro union_relations(relations, where=none, column_override=none, include=[], exclude=[], source_column_name='_dbt_source_relation') -%}
+    {{ return(adapter.dispatch('union_relations', 'dbt_utils')(relations, where, column_override, include, exclude, source_column_name)) }}
 {% endmacro %}
 
-{%- macro default__union_relations(relations, column_override=none, include=[], exclude=[], source_column_name='_dbt_source_relation') -%}
+{%- macro default__union_relations(relations, where=none, column_override=none, include=[], exclude=[], source_column_name='_dbt_source_relation') -%}
```
[the remark this answers is not in the document]
Comment (LewisDavies) [reply macros/sql/union.sql:5]: Good call - there wasn't a particular reason I put it first so I'll move it to the end

**Comment by dbeatty10** [comment]
> @dbeatty10 Thanks for the feedback! I've made the changes and tests are passing.
> 
> I rebased locally onto `next/minor`, pushed my changes, and changed the target branch. Should I go ahead and update `CHANGELOG.md` too?

Sorry for the dramatic delay in conversation @LewisDavies 

Could I bother you for two more changes and then we can merge this in?
1. Rebase to `main` -- we just switched to all pull requests basing off main for simplicity sake
2. Yes, go ahead and update `CHANGELOG.md` within the "Unreleased" section.

If you don't have the time, just let me know and I can do those two steps too.

Thank you so much for your detailed solution on this!

**Comment by LewisDavies** [comment]
@dbeatty10 No worries, the changes are all done. Since I rebased off main I updated the target branch too.

With all my testing and rebasing the commits were an absolute mess, so I cleaned them up and force pushed again. Should be good to go now!

**Review by dbeatty10 — APPROVED** [review]
Looks awesome, @LewisDavies ! This rocks 🎸

**Pull request merged**